pandas/core/indexes/base.py

```diff
@@ -4872,7 +4902,23 @@ def _get_engine_target(self) -> ArrayLike:
             # GH#45652 much more performant than ExtensionEngine
             return vals._ndarray
         if type(self) is Index and isinstance(self._values, ExtensionArray):
-            # TODO(ExtensionIndex): remove special-case, just use self._values
+            from pandas.core.arrays import BaseMaskedArray
+
+            if not isinstance(self._values, BaseMaskedArray):
+                # TODO(ExtensionIndex): remove special-case, just use self._values
+                return self._values.astype(object)
+        return vals
+
+    def _get_join_target(self) -> ArrayLike:
+        """
+        Get the ndarray or ExtensionArray that we can pass to the IndexEngine
+        constructor.
+        """
+        vals = self._values
+        if isinstance(vals, StringArray):
+            # GH#45652 much more performant than ExtensionEngine
+            return vals._ndarray
+        if type(self) is Index and isinstance(self._values, ExtensionArray):
             return self._values.astype(object)
```

Comment:
IIRC _can_use_libjoin will make it so we don't go down this path

Reply:
Yes, but we can not use this in one case, because the no use lib join path does not work there for non interval, so we have to go through object and end up here

Reply:
line 4444